### PR TITLE
Workflow to auto deploy to staging

### PR DESCRIPTION
**Pull request by asteel-gsa**
Reference: This was discussed during the engineering sync about creating automation for deploying to staging more frequently.

**Considerations:**
Testing: [Repo](https://github.com/asteel-gsa/workflow)
Set a nightly schedule for 4:30a UTC or 12:30a EST. This may or may not work as expected. There have been some rumblings about schedules not working, or at least not running at the time set. 
```yaml
on:
    schedule:
    - cron: "30 4 * * *"
```

I tested this instead with:
```yaml
on:
  workflow_dispatch:
```

Because I used PAT when testing this, consideration will need to happen with the `${{ secrets.GITHUB_TOKEN }}`. Per [this](https://github.com/marketplace/actions/auto-approve) - By default, this will use the [automatic GitHub token](https://docs.github.com/en/actions/security-guides/automatic-token-authentication) that's provided to the workflow. This means the approval will come from the "github-actions" bot user. Make sure you enable the `pull-requests: write` permission in your workflow. I think that `pr_reviewer` can be removed, and `GitHub-Actions` bot will be the approver. 

Setting the PR reviewer:
```yaml
pr_reviewer: "fac-admins"
pr_allow_empty: false
github_token: ${{ secrets.GITHUB_TOKEN }}
```
Auto Approving:
```yaml
with:
  pull-request-number: ${{ needs.pull-request.outputs.pr_number }}
  github-token: ${{ secrets.GITHUB_TOKEN }}
 ```

One final note, is that there may be a race condition in place, that we will need to test. This may occur when the PR is created. That will have to be answered by a maintainer, as there needs to be confirmation that `if a PR has been approved, it cannot be merged until the checks pass from triggers-PR.yml`. Branch protection rules on `prod` will also need to be in place to ensure that no PR can be merged if there are conflicts.

**Outputs:**
`Working deployment, without merge conflicts:`
image

`Working run, if there have been no commits within 24 hours:`
image

`What PRs will look like:`
image

CC: @jadudm @JeanMarie-TTS @mogul

### CODE REVIEW

**Comment by JeanMarie-TTS** [comment]
@asteel-gsa , why do we need to create a PR to merge main into prod and then auto=approve it? Can't we just do the merge directly? 
Many of the PR tests seem quite redundant as they all run in a GH runner.
We should invoke the Terraform actions for the staging environment, not for the dev environment.
If there are conflicts, etc, we should just stop the job and have a manual intervention.

**Comment by asteel-gsa** [comment]
> @asteel-gsa , why do we need to create a PR to merge main into prod and then auto=approve it? Can't we just do the merge directly? 

Because of this: [automerge](https://github.com/peter-evans/enable-pull-request-automerge) and even if we use the cli we still need [a pr](https://cli.github.com/manual/gh_pr_merge) to use them.  [Also, that is our promotion to upper environments approach](https://github.com/GSA-TTS/FAC/blob/main/docs/deploying.md)

#### Deploying to `staging`

1.  Create a pull request using `main` targeting `prod`
    *   In the GitHub interface, this will mean setting **base** to `prod` and **compare** to `main`.
    *   The title of the pull request should start with `[YYYY-MM-DD]` and indicate that it’s a merge from `main` to `prod`, for example `2023-04-03 main -> prod`.
2.  Request reviewers.
3.  Wait for the checks to pass.
4.  If checks fail, or if reviewers request changes, something has gone awry. Investigation and/or starting over from a branch and making a PR against `main` may be required.
5.  A member of the [FAC admins team](https://github.com/orgs/GSA-TTS/teams/fac-admins) has to merge the pull request.
    *   Only use “merge pull request” here—_do not use “squash and merge” or “rebase and merge”_ as that will cause `main` and `prod` diverge in Git history terms.
5.  Verify that the deploy steps all passed.
6.  After deployment, the changes should be on https://fac-staging.app.cloud.gov/.

I think it gives us a safer approach, and a way to resolve issues moving from `dev -> staging` without treating `staging` like a `second dev`. If we want a second dev, then we should probably consider a new `cf space` and treat it as a `pre-prod` environment. That being said, I think based on the promotion methodology, `staging` already is a `pre-prod` and not like a `dev sandbox`

> Many of the PR tests seem quite redundant as they all run in a GH runner. 

When a PR is created and merged into main, we already know if it is operational in main. However, given we want to automate our deploys. 

It largely depends on how often we want to promote to staging for #1137 and if we want to use an auto pr/approve/merge for other environments. If we have other environments, we would want their own terraform, and will likely execute in a similar way that we are now. I don't know if these deployments will occur without a Fac-Admin review, so if Github Actions Bot `${{ secrets.GITHUB_TOKEN }}` can create the PR and merge it, that would mean either we figure out 
```yaml
on:
  schedule:
  - cron '<>'
```
and allow it to merge every night, or if we want developers to  call the workflow with
```yaml
on:
  workflow_dispatch:
```
There is some information [here](https://github.com/peter-evans/enable-pull-request-automerge#create-pull-request-example) for the usage of the workflow.

We don't want to promote to staging without running tests, yes, but that is where @eazapata23 may have some insight for replacing [Zap Scan](https://github.com/GSA-TTS/FAC/pull/1135/files#diff-8426d1519269f6a8a33ea51228e9d583782963d08abeebeef2f88f5b298e6b57R8-R13) with this:
```
JeanMarie (OROS, he/him, CA ) [10:23 AM]
Not zap scan as we run it right after each deployment already.
If you are implenting this, you can start by using zap scan as a placeholder. We should replace it with actual tests that make sure all components are working together - are we able to access via [data.gov](http://data.gov/), is the app able to talk to [sam.gov](http://sam.gov/), can we upload an excel file, can the app send an email, etc.
```

> We should invoke the Terraform actions for the staging environment, not for the dev environment. If there are conflicts, etc, we should just stop the job and have a manual intervention.

We should still invoke [this terraform plan here](https://github.com/GSA-TTS/FAC/blob/2ead61dff036c21299a528d03d70c22f39cdfef1/.github/workflows/triggers-PR.yml#L43-L49) and [terraform apply here](https://github.com/GSA-TTS/FAC/blob/2ead61dff036c21299a528d03d70c22f39cdfef1/.github/workflows/triggers-staging.yml#L20-L28). Plus, if the terraform plan, and terraform apply are different for any reason, they will be caught in the PR merging into `prod`, so the apply won't occur.

@JeanMarie-TTS

**Comment by asteel-gsa** [comment]
Added Git Sha to `pr_title` and `pr_body` [example](https://github.com/asteel-gsa/workflow/pull/36)
Which now gives us accurate tracking of the commit hash that is being merged

image

**Comment by asteel-gsa** [comment]
- `github_token: ${{ secrets.PR_CREATOR_TOKEN }}` will have to be set in secrets to be different than runtime `${{ secrets.GITHUB_TOKEN }}`

- Uses new auto merge, that supports waiting for PR checks
```
# 25 Minutes of retries before automerge fails
          MERGE_RETRIES: 50
          MERGE_RETRY_SLEEP: 30000
```

- Adds a "automerge" tag for the script
- Will probably need to add `Tests for PRs / test-and-lint / test (pull_request)` to the branch requirements like i did here for testing to make sure the PR will not automerge until checks have passed
image

File: .github/workflows/auto-create-pr.yml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - id: automerge
+        name: automerge
+        uses: pascalgn/automerge-action@v0.15.6
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          MERGE_RETRIES: 50
+          MERGE_RETRY_SLEEP: 30000
+          MERGE_METHOD: merge
```
Comment (asteel-gsa) [inline .github/workflows/auto-create-pr.yml:90]: Use [Github Automerge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/automatically-merging-a-pull-request)

**Comment by asteel-gsa** [comment]
> * `github_token: ${{ secrets.PR_CREATOR_TOKEN }}` will have to be set in secrets to be different than runtime `${{ secrets.GITHUB_TOKEN }}`

The following will need a resolution before this can be implemented. If we want a "Service Account" or not

**Review by JeanMarie-TTS — APPROVED** [review]

File: .github/workflows/auto-create-pr.yml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+          source_branch: "main"
+          destination_branch: "prod"
+          pr_title: "${{ steps.date.outputs.date }} ${{ steps.git-short.outputs.short_sha }} main -> prod"
+          pr_label: "automerge"
+          pr_body: |
+            This is an auto-generated PR to merge main into prod for a staging release for ${{ steps.date.outputs.date }}
+            Last commit being merged: ${{ steps.git-short.outputs.short_sha }}
+          pr_reviewer: "fac-admins"
+          pr_allow_empty: false
+          github_token: ${{ secrets.FAC_BOT_TOKEN }}
```
[the remark this answers is not in the document]
Comment (asteel-gsa) [reply .github/workflows/auto-create-pr.yml]: latest change removed this:

```yml
          pr_reviewer: "fac-admins"
          pr_allow_empty: false
          github_token: ${{ secrets.GITHUB_TOKEN }}
```

File: .github/workflows/auto-create-pr.yml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  auto-approve:
+    name: Auto Approve Created PR
+    needs: [pull-request]
```
[the remark this answers is not in the document]
Comment (asteel-gsa) [reply .github/workflows/auto-create-pr.yml]: outdated and removed for initial work. 

Github Branch Protections will not allow a merge to occur, if the status checks have not passed. the pr can be auto approved without the status checks, but cannot be automerged if the status checks do not pass

**Pull request merged**